### PR TITLE
bigtable: Add token timeout backup check

### PR DESCRIPTION
**Pull request by brandon-j-roberts**
#### Problem

Some Solana archival nodes have been disconnecting from bigtable. This causes headache for node providers, since we need to restart these nodes once they disconnect. This issue also causes 30-90 minutes of downtime for the node user, since we need to restart the node to regain access to bigtable.

This issue was thought to be resolved by [this PR](https://github.com/solana-labs/solana/pull/28728), however some of our archival nodes which are receiving a lot of traffic and load are still disconnecting from bigtable. After further research, I have found that sometimes the timeout will neither return or timeout while the `goauth::get_token` async function is running inside the `get_token` method. This then causes `refresh_active` to never be set to false, so future retries to refresh the auth token never go through.

Related Issue : https://github.com/solana-labs/solana/issues/29692

#### Summary of Changes

- Added a check to see if the get token / refresh has been active for 2 times the timeout threshold, and if so, it sets `refesh_active` back to false. This will allow the token to attempt a refresh again next time it is used.

#### Testing

Our team has built the solana 1.13.5 client with this change and applied it to our 4 troubled archival nodes 3 weeks ago. They haven't had any archival disconnects since.
Also, after monitoring the logs we have found that this change has prevented 20+ disconnects, due to getting 20+ `Token refresh timeout failed to timeout!` messages.
Examples:
`[2023-01-10T02:14:41.281879955Z WARN  solana_storage_bigtable::access_token] Token refresh timeout failed to timeout!`
`[2023-01-08T21:41:08.355265909Z WARN  solana_storage_bigtable::access_token] Token refresh timeout failed to timeout!`
`[2023-01-06T07:21:08.327848294Z WARN  solana_storage_bigtable::access_token] Token refresh timeout failed to timeout!`

### CODE REVIEW

File: storage-bigtable/src/access_token.rs [hunk 1]
```diff
@@ -125,6 +136,9 @@ impl AccessToken {
                 warn!("Token refresh timeout")
             }
         }
+
+        info!("Token refresh Complete!");
```
Comment (mvines) [inline storage-bigtable/src/access_token.rs:140]:
```suggestion
        info!("Token refresh complete");
```

File: storage-bigtable/src/access_token.rs [hunk 2]
```diff
@@ -105,13 +109,20 @@ impl AccessToken {
                 .compare_and_swap(false, true, Ordering::Relaxed)
             {
                 // Refresh already pending
+                let token_refresh_time = self.token_refresh_start_time.load(Ordering::SeqCst);
```
Comment (mvines) [inline storage-bigtable/src/access_token.rs:112]: Any reason why you didn't use `Ordering::Relaxed` here?

Comment (mvines) [reply storage-bigtable/src/access_token.rs:112]: Also I'd prefer if `token_refresh_time` was renamed to `token_refresh_start_time` for (arguable) improved code clarity

**Review by mvines — COMMENTED** [review]
Thanks for this PR!

Thanks 

>  I have found that sometimes the timeout will neither return or timeout while the `goauth::get_token` async function is running inside the get_token method

Have you researched why this is happening within `goauth::get_token`? This is somewhat concerning because if the function is in fact not returning, then a thread was just blocked potentially permanently. In such a case, this PR then calls `goauth::get_token` on another thread potentially blocking it as well. If this process continues we could expect to eventually see all threads blocked.

**Comment by RdeWilde** [comment]
@mvines at first I was not really a fan about this kind of hacky solution, but I came around sort of. I'm been investigating the same issue and found that the underlying connection is stuck without any obvious reason or returned error. This presumably happens in hyper::client which is leverage by reqwest. I've updated to the latest dependencies, set timeouts, disabled connection pooling, add additional logging and enabled debug/trace logs, but still it is happening.

This makes me realize we should not rely on 3rd parties to handle thread. They can always contain bugs (now or later) and this should not be able to affect the functionality of solana's software.

I hear your objection with fixing the underlying problem and threads filling up, but as I see it:
- long term, fix the underlying problem (or assign resources right now).
- kill the whole thing on timeout (which isn't happening right now).
- use this "workaround" for near term
This issue is seriously damaging reputation of Solana's software and ability to fix bugs quickly, with bigger companies. This bug has been existing for over a year and should be resolved one way or the other quickly. It's alerting our team out of bed at random times, having to tell the node needs to restart again (which takes like an hour) and tell them again there's no solution yet.

If interested I'd be happy to share any results from trying to debug this, other than I already provided above. Can we please fix this? Currently it's really a pain in the ass.

File: storage-bigtable/src/access_token.rs [hunk 3]
```diff
@@ -105,13 +109,20 @@ impl AccessToken {
...
+                if token_refresh_time != 0 && token_refresh_time + (self.get_token_timeout_seconds * 2) < token_r.1.elapsed().as_secs() {
+                    warn!("Token refresh timeout failed to timeout!");
```
Comment (RdeWilde) [inline storage-bigtable/src/access_token.rs:114]: This message is confusing. I would suggest remote "to timeout" or make it more descriptive.

File: storage-bigtable/src/access_token.rs [hunk 4]
```diff
@@ -57,6 +59,8 @@ impl AccessToken {
                 credentials,
                 scope,
                 token,
+                token_refresh_start_time: Arc::new(AtomicU64::new(0)),
+                get_token_timeout_seconds: 5,
```
Comment (RdeWilde) [inline storage-bigtable/src/access_token.rs:63]: I think this is a very optimistic timeout and personally have increase this a bit.

**Comment by joeaba** [comment]
Reopening since I've talked to some folks that wanna take over this
https://github.com/solana-labs/solana/issues/29692

**Comment by nyetwurk** [comment]
Please don't let this languish.

**Comment by PeaStew** [comment]
Still doesn't work

**Pull request closed without merging**